### ISSUE TITLE
`_gels` fails for fat A matrix

### ISSUE THREAD

**Issue by pstjohn** [scipy/scipy#8329]
Replicating the tests in `scipy/linalg/tests/test_lapack.py` for a fat A matrix, `dgels` fails with `ValueError: Internal work array size computation failed: -8`. 

```python
m = 10
n = 20

a1 = np.ones((m, n))
b1 = np.ones(m)
gels, gels_lwork, geqrf = get_lapack_funcs(
        ('gels', 'gels_lwork', 'geqrf'), (a1, b1))

m, n = a1.shape
if len(b1.shape) == 2:
    nrhs = b1.shape[1]
else:
    nrhs = 1

# Request of sizes
lwork = _compute_lwork(gels_lwork, m, n, nrhs)
```

This is probably because LAPACK [asks](http://www.netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga225c8efde208eaf246882df48e590eac.html) for `LDB >= MAX(1,M,N)`, while this line 
https://github.com/scipy/scipy/blob/ad4a557eadbc5d940f33f00a9dd9729ad7184fd3/scipy/linalg/flapack.pyf.src#L1481
sets ldb to M, which might be less than N.

**Comment by pstjohn** [comment]
Just changing `ldb` wont be enough for a fix, we'll also need to handle the changing size of the `b` vector. Since it's also used for the x output, the input shape(m) isn't necessarily the output shape (n).

**Comment by ilayn** [comment]
The failing function is the lwork query so it shouldn't fail with -8 but something else. ldb is always the leading dimension of b so it is not related to this failure.

**Comment by pstjohn** [comment]
the lwork query just calls the lapack function with -1 for lwork. If you call gels(..., lwork=something) you get the same response. Lapack returns a negative number when there is an error with an input argument, in this case the 8th input argument, ldb.

While yes, ldb is the leading dimension of b, b in this case should be 'expanded' to the expected dimension of x: since lapack uses B as both the input right-hand-side and the output solution.

It looks like this is done with the other `gels*' algorithms, but not this one.

**Comment by ilayn** [comment]
Yes I am aware of these. What I mean is that the lwork query when supplied with -1 should not give -8 because that's what it is complaining about. There is something else going on, I'll try to check in a few hours if someone else doesn't chime in.

**Comment by pstjohn** [comment]
Fair enough, let me know what you find. I was assuming the -8 came from a preprocessing directive in `gels`, which would be run regardless of whether `gels` was called for a solution or to calculate the optimal `lwork`.

**Comment by ilayn** [comment]
From a quick look it's indeed not gels but gels_work ldb parameter

https://github.com/scipy/scipy/blob/ad4a557eadbc5d940f33f00a9dd9729ad7184fd3/scipy/linalg/flapack.pyf.src#L1506

**Comment by ilayn** [comment]
you can check line 243 down the page http://www.netlib.org/lapack/explore-html/d7/d3b/group__double_g_esolve_ga225c8efde208eaf246882df48e590eac.html#ga225c8efde208eaf246882df48e590eac

**Comment by ilayn** [comment]
@pstjohn Whenever you have the chance can you test the PR linked above?

**Comment by matthew-brett** [comment]
I noticed a new failure in the wheel builds:

```
../venv/lib/python3.5/site-packages/numpy/testing/utils.py:665: AssertionError
=============================== warnings summary ===============================
scipy/linalg/tests/test_basic.py::TestLstsq::()::test_simple_overdet
  /Users/travis/build/MacPython/scipy-wheels/venv/lib/python3.5/site-packages/scipy/linalg/tests/test_basic.py:988: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
    overwrite_b=overwrite)
scipy/linalg/tests/test_basic.py::TestLstsq::()::test_simple_exact
  /Users/travis/build/MacPython/scipy-wheels/venv/lib/python3.5/site-packages/scipy/linalg/tests/test_basic.py:956: RuntimeWarning: internal gelsd driver lwork query error, required iwork dimension not returned. This is likely the result of LAPACK bug 0038, fixed in LAPACK 3.2.2 (released July 21, 2010). Falling back to 'gelss' driver.
    overwrite_b=overwrite)
```

https://travis-ci.org/MacPython/scipy-wheels/jobs/334138245

Is this related?

**Comment by pv** [comment]
Those warnings are expected on OSX and indicate accelerate bug. (The actual test failure is in test_hdmedian, which hopefully is addressed by gh-8334)
